### PR TITLE
Introduce a tracker API and integrate Aimstack tracker

### PR DESCRIPTION
**Pull request by dushyantbehl**
# What does this PR do?

This pull request adds a tracker API to llama recipes and integrates Aimstack https://aimstack.io/
The goal of tracker API is to create a standard template to attach any other experiment trackers.

With aim integration there is a dependency on the `aim` package which needs to be installed separately before using this.

Fixes # (issue)

Relates to issue https://github.com/facebookresearch/llama-recipes/issues/225

#Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
      https://github.com/facebookresearch/llama-recipes/issues/225
- [x] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

**Comment by dushyantbehl** [comment]
Opening up the PR for initial review. Feature has to be fully tested and unit test and documentation fixed.

**Comment by dushyantbehl** [comment]
@HamidShojanazeri @nairbv opening this PR for feedback. Sorry it has been long and I have been occupied with other things. 
I have added a tracker API which can be used to add support for any tracker like WandB / Aim along with Aim integration.

File: docs/experiment_tracking.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Experiment tracking in llama recipes. 
+
+Llama recipes has an optional support for tracking experiments using experiment tracker modules like Aimstack ().
+
+The experiment tracker can be enabled by passing an additional argument to the llama recipes code which loads and
+tracks all the training and evaluation metrics in the tracker along with the system metrics being tracker by these trackers.
+
+The addition argument can be passed as -
```
Comment (nairbv) [inline docs/experiment_tracking.md:8]: missing the argument? looks like the below might be a placeholder?

File: requirements.txt [hunk 2]
```diff
@@ -12,4 +12,5 @@ transformers>=4.31.0
 sentencepiece
 py7zr
 scipy
-optimum
\ No newline at end of file
+optimum
+aim=3.7.11
```
Comment (nairbv) [inline requirements.txt:16]: probably want to make this optional. I think we're going to want to ensure that users that don't have (don't want) aim installed should still be able to use llama-recipes.

File: src/llama_recipes/utils/train_utils.py [hunk 3]
```diff
@@ -66,6 +67,11 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
     checkpoint_times = []
     results = {}
     best_val_loss = float("inf")
+
+    track_metrics = True
+    if train_config.enable_fsdp and rank != 0:
+        track_metrics = False
```
Comment (nairbv) [inline src/llama_recipes/utils/train_utils.py:73]: just `track_metrics = ....` in one line is a bit cleaner. something like `track_metrics = not fsdp or rank == 0` I think would be equivalent.

File: src/llama_recipes/utils/train_utils.py [hunk 4]
```diff
@@ -114,24 +123,27 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+
+        if tracker is not None and track_metrics:
+            tracker.track(epoch_end_time , name='epoch_times', stage='train')
+            tracker.track(train_epoch_loss , name='train_loss', stage='train')
+            tracker.track(train_perplexity , name='train_perplexity', stage='train')
+            tracker.track(memtrace.peak, name="Max CUDA Mem Allocated(GB)", stage='train')
+            tracker.track(memtrace.max_reserved, "Max CUDA Mem Reserved(GB)", stage='train')
+            tracker.track(memtrace.peak_active_gb, "Peak active CUDA Mem(GB)", stage='train')
+            tracker.track(memtrace.cuda_malloc_retires, "Cuda Malloc retires", stage='train')
+            tracker.track(memtrace.cpu_peaked + memtrace.cpu_begin, "CPU Total Peak Mem Max(GB)", stage='train')
```
Comment (nairbv) [inline src/llama_recipes/utils/train_utils.py:142]: It looks a bit awkward having two very similar pieces of code logging the same info two different ways.

Would it make sense to combine print logging above with the 'tracker' here somehow? e.g. maybe an alternate default implementation of tracker=PrintTracker that just prints. Or maybe the set of things to be logged can be passed to both the tracker and print statements in a loop.

File: src/llama_recipes/utils/train_utils.py [hunk 5]
```diff
@@ -179,17 +191,14 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+            if tracker is not None and track_metrics:
+                tracker.track(best_val_loss , name='best_val_loss', stage='validation')
+                tracker.track(eval_epoch_loss , name='eval_epoch_loss', stage='validation')
             val_loss.append(best_val_loss)
             val_prep.append(eval_ppl)
-        if train_config.enable_fsdp:
-            if rank==0:
-                print(f"Epoch {epoch+1}: train_perplexity={train_perplexity:.4f}, train_epoch_loss={train_epoch_loss:.4f}, epoch time {epoch_end_time}s")
-        else:
+        if track_metrics:
```
Comment (nairbv) [inline src/llama_recipes/utils/train_utils.py:201]: it looks like this change is removing the logging of epoch loss and time in the fsdp case?

File: src/llama_recipes/utils/train_utils.py [hunk 6]
```diff
@@ -98,10 +104,13 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
                         optimizer.step()
                         optimizer.zero_grad()
                         pbar.update(1)
+                if tracker is not None and track_metrics:
+                    tracker.track(loss, name='loss', stage='train')
```
Comment (nairbv) [inline src/llama_recipes/utils/train_utils.py:108]: should we be gathering loss from all ranks in the fsdp case?

Comment (nairbv) [reply src/llama_recipes/utils/train_utils.py:108]: also, can we track throughput (tokens per second and/or tokens per step) during training?

**Comment by nairbv** [comment]
looks like there are conflicts with main that will have to be resolved too

File: src/llama_recipes/utils/train_utils.py (hunk 5 above)
Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py:201]: The original code checks if fsdp is enabled and if so prints only on rank 0. if fsdp is not enabled it prints anyway.

`track_metrics` is used as a short cut to check this condition. It is set as `not (train_config.enable_fsdp and rank != 0)` to mimic this condition and remove a duplicate print statement.

File: src/llama_recipes/utils/train_utils.py (hunk 4 above)
Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py:142]: Our team's patch which fixes GPU metrics tracking is merged in Aim now. We do not need extra memory tracking I think.

Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py:142]: Keeping the interface simple and updated the code. Any extra can be done as additonal patch in my opinion.

File: src/llama_recipes/utils/train_utils.py (hunk 6 above)
Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py:108]: The original code tracked only on rank 0 in fsdp. 
If we need to track on all ranks in Aim, we can do that.

File: src/llama_recipes/utils/train_utils.py (hunk 3 above)
Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py:73]: done

File: src/llama_recipes/utils/train_utils.py (hunk 5 above)
Comment (nairbv) [reply src/llama_recipes/utils/train_utils.py:194]: ah maybe I misread the condition the first time

File: src/llama_recipes/utils/train_utils.py [hunk 7]
```diff
@@ -207,6 +209,17 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
     results["avg_epoch_time"] = avg_epoch_time
     results["avg_checkpoint_time"] = avg_checkpoint_time
 
+    if tracker is not None and track_metrics:
+        tracker.track(avg_train_loss , name='avg_train_loss', context={'subset':'experiment'})
```
Comment (nairbv) [inline src/llama_recipes/utils/train_utils.py:213]: wrong parameter names? let's make sure we test this to the end of an epoch. we can truncate a dataset so it runs to completion faster.

Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py]: Fixed. Tested and updated the documentation.

File: docs/experiment_tracking.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
+# Experiment tracking in llama recipes. 
+
+Llama recipes has an optional support for tracking experiments using experiment tracker modules like Aimstack or WandB (Currently not supported).
```
Comment (nairbv) [inline docs/experiment_tracking.md:3]: nit:
I'd drop 'an' in 'an optional'

I think "Currently only Aimstack is supported" might be clearer, "not supported' doesn't make it clear what is/isn't.

File: docs/experiment_tracking.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+Current version of the code supports Aimstack tracking via the trackers module.
+
+One needs to install Aimstack separately if need to use aimstack as a tracker.
+
+### Installation
+
+Aim can be installed by running
+
+```
+$ pip install aim==3.17.5
```
Comment (nairbv) [inline docs/experiment_tracking.md:30]: do we require a specific version or can we put `>=` ?

File: docs/experiment_tracking.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+### Aim setup
+
+Aimstack is a self hosted open source component so a small setup is needed before running Aimstack.
```
Comment (nairbv) [inline docs/experiment_tracking.md:35]: nit: s/a small/some/

File: docs/experiment_tracking.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+           --model_name $MODEL_NAME --output_dir $OUTPUT_DIR \
+           --use_peft --peft_method lora --pure_bf16 \
+           --enable_fsdp --use_fast_kernels --num_epochs 1 --batch_size_training 1\
+           --tracker aim --aim_config.repo $AIMSTACK_DB \
+           --aim_config.experiment llama7b-finetuning
+```
+
+### Example screenshots.
+
+This is how the runs would look like,
```
Comment (nairbv) [inline docs/experiment_tracking.md:80]: nit: either "how the runs look" or "what the runs look like" not "how the runs look like"

When referring to a code snippet or image, it might also be better to be consistent in terms of punctuation. some end in a comma, some a period, some nothing. Maybe a colon would be better?

File: src/llama_recipes/utils/train_utils.py [hunk 12]
```diff
@@ -207,6 +209,17 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+    if tracker is not None and track_metrics:
+        tracker.track(avg_train_loss , name='avg_train_loss', stage={'subset':'experiment'})
+        tracker.track(avg_train_prep, name='avg_train_prep', stage={'subset':'experiment'})
+
+        if train_config.run_validation:
+            tracker.track(avg_eval_prep , name='avg_eval_prep', stage={'subset':'validation'})
+            tracker.track(avg_eval_loss, name='avg_eval_loss', stage={'subset':'validation'})
+
+        tracker.track(avg_epoch_time , name='avg_epoch_time', stage={'subset':'experiment'})
+        tracker.track(avg_checkpoint_time , name='avg_checkpoint_time', stage={'subset':'experiment'})
```
[the remark this answers is not in the document]
Comment (dushyantbehl) [reply src/llama_recipes/utils/train_utils.py]: yes this was fixed. thanks.

File: docs/experiment_tracking.md (hunk 9 above)
Comment (dushyantbehl) [reply docs/experiment_tracking.md:30]: Aimstack sdk is changing across versions so I am putting the version which I have found to be stable.

File: src/llama_recipes/configs/trackers.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.
+
+from dataclasses import dataclass
+
+@dataclass
+class aim_config:
+    experiment: str = "llama-recipes"
+    # This is either the location of a locallly accessible directory example `.run`
```
Comment (nairbv) [inline src/llama_recipes/configs/trackers.py:9]: nit: seems like there's some punctuation missing?

File: src/llama_recipes/configs/trackers.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+    # or can be the location of a remote repository which is hosted on a server.
+    # if remote_server_ip or remote_server_port is set then the repo will be set to
+    # remote aim repo, else the directory specified by `repo` takes precedence.
+    # The directory defaults to None which means `.run` in aim.
```
Comment (nairbv) [inline src/llama_recipes/configs/trackers.py:13]: is the default `~/.run` or really just `.run` ?

**Review by nairbv — APPROVED** [review]
Some minor comments but overall LGTM.

We probably need to have someone from Meta who has committed to this repo take a look before merging. @HamidShojanazeri / @WuhanMonkey  / @jeffxtang  ?

**Comment by dushyantbehl** [comment]
@nairbv @HamidShojanazeri @chauhang I noticed a spell check was failing for this PR which is fixed now. Please take a look and let me know if anything else needs to be fixed.
